Refactor postEditor#insertPost to handle more situations

bantic opened the pull request:
Ready to start getting feedback on this. There are some todos and fixmes I intend to address before this should be merged, though.

  * refactor postEditor#insertPost to handle a larger number of situations
  * fix bug in post#cloneRange when range starts in list item
  * add assert.isPostSimilar, assert.isRenderTreeEqual, assert.isPositionEqual
  * add test for ListSection#clone

Fix #249 #259

mixonic commented on src/js/editor/editor.js:
```diff
@@ -754,8 +754,7 @@ class Editor {
 
     this.run(postEditor => {
-      let nextPosition = postEditor.insertPost(position, pastedPost);
-      postEditor.setRange(new Range(nextPosition));
```
Can this be left outside `insertPost`?

We discussed the range stuff in detail but I don't remember if we talked about moving cursor changes into `postEditor` methods. Am I forgetting a reason this must be done?

bantic replied:
I'm not sure what you mean by "outside", but this can still use the position-returning semantics...I switched it back to the way it was. `postEditor.insertPost` now returns a `Position`

mixonic replied:
Cool. I'm not sure if we had the discussion or not. I know having mutation methods with the side effect of changing cursor position is something we've avoided until now. However the arguments against it are fewer now that we have `setRange`, a safe way to set the position and read it back without the DOM.

mixonic commented on src/js/editor/post.js:
```diff
@@ -720,6 +721,30 @@ class PostEditor {
     return this.moveSectionBefore(collection, renderedSection, beforeSection);
   }
 
+  // TODO write tests for this
```
Notable `TODO` ;-)

mixonic commented on src/js/editor/post.js:
```diff
@@ -720,6 +721,30 @@ class PostEditor {
+
+    let edit = section.splitMarkerAtOffset2(offset);
+    edit.removed.forEach(marker => this._scheduleForRemoval(marker));
+    edit.added.forEach(marker => this._markDirty(marker));
```
Added things should not need to be marked dirty, they are already dirty. `this._markDirty(section)` as you have below is sufficient.

mixonic commented on src/js/editor/post.js:
```diff
@@ -720,6 +721,30 @@ class PostEditor {
+    this._coalesceMarkers(section);
+    this._markDirty(section);
+
+    position.offset = offset;
```
Can this return a new `Position` instead of mutating the one passed in?

```js
return new Position(position.section, offset);
```

mixonic commented on src/js/editor/post/post-inserter.js:
```diff
@@ -0,0 +1,303 @@
+  _breakMarkerableAtCursor() {
+    let [pre, post] = this.cursorSection.splitAtPosition(this.cursorPosition)
+                          .map(s => s.clone());
+    this.postEditor._coalesceMarkers(pre);
```
Maybe `coalesceMarkers` should be public?

bantic replied:
I think `_coalesceMarkers` is an implementation detail that no consumer should have a need to use. It was in use here because this code breaks a section without going through the postEditor. I changed it to use `postEditor#splitSections` instead

mixonic commented on src/js/models/_markerable.js:
```diff
@@ -11,7 +12,11 @@ export default class Markerable extends Section {
+        assert(`Cannot insert non-marker into markerable (was: ${m.type})`,
+               m.isMarker);
+        m.section = m.parent = this;
+      },
```
This is one of the reasons I dislike the implicit return on `=>` without `{`. Was the return value important here and this refactor is in error? It is difficult to know when reading.

But I know from experience it is not in error. Just a note to share.

mixonic commented on src/js/models/_markerable.js:
```diff
@@ -127,11 +132,67 @@ export default class Markerable extends Section {
     return edit;
   }
 
+  // splits the marker that contains (or is before) the given offset, at the offset
```
period at the end of this line?

mixonic commented on src/js/models/_markerable.js:
```diff
@@ -127,11 +132,67 @@ export default class Markerable extends Section {
+  // when splitting at the start.
+  // This method will modify the section such that `markerBeforeOffset` always
+  // has a valid return value if given the same offset
+  splitMarkerAtOffset2(sectionOffset) {
```
Should this be in `postEditor`? Is there a plan to resolve the names?

bantic replied:
yeah, I intend to resolve the names

mixonic commented on src/js/models/marker.js:
```diff
@@ -116,6 +117,23 @@ const Marker = class Marker extends LinkedItem {
     return markers;
   }
 
+  // @return 2 markers either or both of which could be blank
```
To be YUIDoc you need to use `*` quotes:

```js
/**
 * @return {Array} Two markers surrounding the offset. One or both may be blank
 */
```

mixonic commented on src/js/models/post.js:
```diff
@@ -217,10 +207,27 @@ export default class Post {
-    this.walkPostSections(range, section => {
+    let sectionParent = post,
+        listParent = null;
+    this.walkLeafSections(range, section => {
```
haha, for a different time but I wonder if this can be structured as an abstract renderer :-)

mixonic commented:
Some pretty great stuff!!!

bantic commented:
@mixonic I've added some additional test coverage for new methods and addressed the first round of feedback. This is ready for a final review now. If all looks good, I'll squash and merge.
